Enhance: Better previews for HDR video (CLOSED, IGNORE)

invario opened the pull request:
Hello, hopefully I did this correctly this time, thank you.

* Resolves: #44881 

## Summary
Previews currently generated for HDR videos result in desaturated/washed out images because of lack of tonemapping. This enhancement detects HDR videos and adds additional switches to the ffmpeg command to generate much improved images.

- ffprobe is required (included with ffmpeg)
- ffmpeg **must** have been built with **libzimg** (--enable-libzimg) for the zscale filter to work. This is platform/package dependent. (e.g. FreeBSD 13 does not. Alpine does. Unsure of others.)
- adds configuration option _**preview_ffmpeg_enable_hdr**_
- adds [optional] configuration option _**preview_ffprobe_path**_
- 
**Before**
![before](https://github.com/user-attachments/assets/1f1dcf20-7226-4b61-b629-5e56116854ac)

**After**
![after](https://github.com/user-attachments/assets/c07edff6-16c5-491d-b5a8-b58d47f1f235)

## TODO

- [x] Perform actual check to see if ffmpeg was built with libzimg. However, this would require another proc_open with command `ffmpeg -v quiet -filters` and searching its output for 'zscale'. I don't think it's necessary as HDR preview generation would be disabled by default. Any user enabling the option should be aware of the requirement before doing so.
- [ ] Update config.php.sample to show new options
- [ ] Need additional testing

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [ ] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

nextcloud-bot committed:
Fix(l10n): Update translations from Transifex

Signed-off-by: Nextcloud bot <bot@nextcloud.com>
Signed-off-by: KT <invario@gmail.com>

invario committed:
Enhance: Better previews for HDR video

Previews currently generated for HDR videos result in desaturated/washed out images because of lack of tonemapping. This enhancement detects HDR videos and adds additional switches to the ffmpeg command to generate much improved images.

Signed-off-by: invario <67800603+invario@users.noreply.github.com>
Signed-off-by: KT <invario@gmail.com>

invario committed:
swapped out comma for period

Signed-off-by: invario <67800603+invario@users.noreply.github.com>
Signed-off-by: KT <invario@gmail.com>

icewind1991 commented:
(disclaimer: my knowledge about ffmpeg/HDR is limited)

Is there a downside to using the hdr options for non-hdr content? It would be better to have hdr previews "just work" without having to do any special config would be better (assuming the ffmpeg has the proper support ofc).

Doing the extra `exec` to feature-detect seems fine to me, since we can also cache it if we're worried about the overhead. (Though the cost of `ffmpeg -v quiet -filters` is probably insignificant compared to the actual preview generation)

invario commented:
> (disclaimer: my knowledge about ffmpeg/HDR is limited)
> 
> Is there a downside to using the hdr options for non-hdr content? It would be better to have hdr previews "just work" without having to do any special config would be better (assuming the ffmpeg has the proper support ofc).
> 
Yes. The downside of using the HDR switches on an SDR video is that previews generated are degraded (extremely dark). That's why there is a need to differentiate. Do you need a sample image?

invario commented:
> I mean always doing the logic that is gated behind `preview_ffmpeg_enable_hdr` currently (so including the check for "smpte2084" or "arib-std-b67").

Ah, sorry, misunderstood. No, there's no downside as long as we confirm somewhere earlier that ffmpeg was built with libzimg and ffprobe is also installed.

icewind1991 commented:
Ok, I would be in favor then of always enabling the behavior and doing the feature detection

invario commented:
> Ok, I would be in favor then of always enabling the behavior and doing the feature detection

I can write the routine to detect, and then have the value of `preview_ffmpeg_enable_hdr` set based on the results (instead of a user config option)? 

Not sure where I should insert the routine for least performance impact. What did you mean by 'cache' the results earlier... I admit I don't know how to go about that.

invario commented:
Hang on actually, just figured out a more elegant way to do this.  ffmpeg/ffprobe already display compile options which is output to stderr by default, unless suppressed by `-v quiet`.  

I can use the same `exec` I used for ffprobe and check its stderr for '--enable-libzimg'.  No need to do a separate check.

invario committed:
added check for libzimg (zscale filter)

- removed preview_ffmpeg_enable_hdr option and instead check ffmpeg build options for libzimg and enable/disable based on that

Signed-off-by: invario <67800603+invario@users.noreply.github.com>
Signed-off-by: KT <invario@gmail.com>

invario commented:
There we go.  Now checks for the presence of libzimg (zscale filter) when running ffprobe on the video.

invario committed:
Fix check for libzimg

Use of str_pos on null would result if ffprobe failed to exec.

Signed-off-by: invario <67800603+invario@users.noreply.github.com>
Signed-off-by: KT <invario@gmail.com>

invario commented:
@st3iny @joshtrichards Hello, I run jnto your names frequently when looking at preview generation related stuff. Would you mind taking a look? TIA!

invario commented:
I would love for this to gain some traction or if I can get some guidance on additional steps that I may be missing. Thank you.

icewind1991 commented:
@invario can you resolve the formatting issues from https://github.com/nextcloud/server/actions/runs/13774792167/job/38566161541?pr=51356

invario committed:
Test commit

Signed-off-by: KT <invario@gmail.com>

invario commented:
Okay, I think I fixed the formatting. Apologies for the improper commit message, did not realize it would push here. (I'm learning!)

invario commented:
Once again, apologies, but I'm making a mess of things. :(

invario commented:
I'm going to do a new PR and close this one out.
